cpp/src/arrow/compute/compute-test.cc

```diff
@@ -236,6 +236,65 @@ TEST_F(TestCast, ToIntDowncastSafe) {
   // underflow
   vector<int32_t> v6 = {0, 1000, 2000, -70000, 0};
   CheckFails<Int32Type>(int32(), v6, is_valid, int16(), options);
+
+  vector<int32_t> v7 = {0, 1000, 2000, -70000, 0};
+  CheckFails<Int32Type>(int32(), v7, is_valid, uint8(), options);
+}
+
+TEST_F(TestCast, IntegerSignedToUnsigned) {
+  CastOptions options;
+  options.allow_int_overflow = false;
+
+  vector<bool> is_valid = {true, false, true, true, true};
+
+  vector<int32_t> v1 = {INT32_MIN, 100, -1, 0, INT32_MAX};
+  // Same width
+  CheckFails<Int32Type>(int32(), v1, is_valid, uint32(), options);
+  // Wider
+  CheckFails<Int32Type>(int32(), v1, is_valid, uint64(), options);
+  // Narrower
+  CheckFails<Int32Type>(int32(), v1, is_valid, uint16(), options);
+
+  options.allow_int_overflow = true;
+
+  vector<uint32_t> v2 = {(uint32_t)INT32_MIN, 100, (uint32_t)-1, 0, (uint32_t)INT32_MAX};
+  CheckCase<Int32Type, int32_t, UInt32Type, uint32_t>(int32(), v1, is_valid, uint32(), v2,
+                                                      options);
+
+  vector<uint64_t> v3 = {(uint64_t)INT32_MIN, 100, (uint64_t)-1, 0, (uint64_t)INT32_MAX};
+  CheckCase<Int32Type, int32_t, UInt64Type, uint64_t>(int32(), v1, is_valid, uint64(), v3,
+                                                      options);
+
+  vector<uint16_t> v4 = {(uint16_t)INT32_MIN, 100, (uint16_t)-1, 0, (uint16_t)INT32_MAX};
+  CheckCase<Int32Type, int32_t, UInt16Type, uint16_t>(int32(), v1, is_valid, uint16(), v4,
+                                                      options);
+}
+
+TEST_F(TestCast, IntegerUnsignedToSigned) {
+  CastOptions options;
+  options.allow_int_overflow = false;
+
+  vector<bool> is_valid = {true, true};
+
+  vector<uint32_t> v1 = {0, UINT32_MAX};
+  // Same width
+  CheckFails<UInt32Type>(uint32(), v1, is_valid, int32(), options);
+  // Narrower
+  CheckFails<UInt32Type>(uint32(), v1, is_valid, int16(), options);
+
+  options.allow_int_overflow = true;
+
+  vector<int32_t> v2 = {0, static_cast<int32_t>(UINT32_MAX)};
+  CheckCase<UInt32Type, uint32_t, Int32Type, int32_t>(uint32(), v1, is_valid, int32(), v2,
+                                                      options);
+
+  vector<int64_t> v3 = {0, static_cast<int64_t>(UINT32_MAX)};
+  CheckCase<UInt32Type, uint32_t, Int64Type, int64_t>(uint32(), v1, is_valid, int64(), v3,
+                                                      options);
+
+  vector<int16_t> v4 = {0, static_cast<int16_t>(UINT16_MAX)};
```

Comment:
INT16_MAX?

Reply:
I'll remove this in the next iteration since I create a `UnsafeVectorCast` helper function, but in this specific case I wanted UINT16_MAX explicitly due to the cast of 0xFFFFFFFF would come 0xFFFF which is not INT16MAX.